packages/fuels-programs/src/contract.rs

```diff
@@ -471,13 +471,53 @@ pub fn get_decoded_output(
     contract_id: Option<&Bech32ContractId>,
     output_param: &ParamType,
 ) -> Result<Token> {
+    let null_contract_id = ContractId::new([0u8; 32]);
     // Multiple returns are handled as one `Tuple` (which has its own `ParamType`)
     let contract_id: ContractId = match contract_id {
         Some(contract_id) => contract_id.into(),
         // During a script execution, the script's contract id is the **null** contract id
-        None => ContractId::new([0u8; 32]),
+        None => null_contract_id,
     };
     let encoded_value = match output_param.get_return_location() {
+        ReturnLocation::ReturnData if output_param.is_vector() => {
+            // If the output of the function is a vector, then there are 2 consecutive ReturnData
+            // receipts. The first one is the one that returns the pointer to the vec struct in the
+            // VM memory, the second one contains the actual vector bytes (that the previous receipt
+            // points to).
+            // We ensure to take the right "first" ReturnData receipt by checking for the
+            // contract_id. There are no receipts in between the two ReturnData receipts because of
+            // the way the scripts are built (the calling script adds a RETD just after the CALL
+            // opcode, see `get_single_call_instructions`).
+            // Find the position of the first corresponding ReturnData receipt
+            let pos = receipts
+                .iter()
+                .position(|receipt| {
+                    matches!(receipt,
+                    Receipt::ReturnData { id, data, .. } if *id == contract_id && !data.is_empty())
+                })
+                .expect("There should be at least a ReturnData receipt");
+            // Get the data contained in the one that immediately follows
+            let return_data_receipt = receipts
+                .get(pos + 1)
+                .expect("There should be at least two receipts");
+            // Its contract id must be null (it comes from a script)
+            if !matches!(return_data_receipt,
+                        Receipt::ReturnData { id, .. } if *id == null_contract_id)
+            {
+                return Err(error!(
+                    InvalidData,
+                    "The next should be a ReturnData receipt with non-empty data and a \
+                    null contract id: {:?}",
+                    return_data_receipt
+                ));
+            };
+            Some(
+                return_data_receipt
+                    .data()
+                    .expect("ReturnData should have data")
+                    .to_vec(),
+            )
+        }
```

Comment:
A functional alternative
Iterate in tuples of two. If you find a pair that satisfies the given conditions, return the data from the second one:
```rust
            let vector_data = receipts
                .iter()
                .tuple_windows()
                .find_map(|(current_receipt, next_receipt)| extract_vec_data(current_receipt, next_receipt, contract_id))
                .cloned()
                .expect("Could not extract vector data");
```
```rust
fn extract_vec_data<'a>(
    current_receipt: &Receipt,
    next_receipt: &'a Receipt,
    contract_id: ContractId,
) -> Option<&'a Vec<u8>> {
    match (current_receipt, next_receipt) {
        (
            Receipt::ReturnData {
                id: first_id,
                data: first_data,
                ..
            },
            Receipt::ReturnData {
                id: second_id,
                data: vec_data,
                ..
            },
        ) if *first_id == contract_id
            && first_data.is_empty()
            && *second_id == ContractId::zeroed() =>
        {
            Some(vec_data)
        }
        _ => None,
    }
}

```

Reply:
I don't prefer this functional approach because IMO it  hinders the readability/understanding you get of the structure of the receipts, and possible debugging steps (you can't really get the same error messages with the functional approach). While implementing `Bytes` especially I really liked having these error messages.